pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -311,20 +334,45 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 
 	cidrStrings := make([]string, 0)
 
-	if len(instance.NetworkInterfaces) == 0 || (len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 0) {
+	// No Pod CIDRs can be allocated to a node, if there are no network interfaces;
+	// or if there is one network interface, but it has neither an IP alias nor an IPv6 address.
+	if len(instance.NetworkInterfaces) == 0 ||
+		(len(instance.NetworkInterfaces) == 1 &&
+			len(instance.NetworkInterfaces[0].AliasIpRanges) == 0 &&
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) == nil) {
+
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated", node.Name)
 	}
 
-	// sets the v1.NodeNetworkUnavailable condition to False
+	// Sets the v1.NodeNetworkUnavailable condition to False.
 	ca.setNetworkCondition(node)
 
-	// nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface with 1 alias IP range.
-	if len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 {
-		cidrStrings = append(cidrStrings, instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange)
-		ipv6Addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0])
-		if ipv6Addr != nil {
-			cidrStrings = append(cidrStrings, ipv6Addr.String())
+	// Nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface
+	// with 1 alias IPv4 range and/or 1 IPv6 address. Multi-network clusters may have 1 interface
+	// with multiple aliases.
+	if len(instance.NetworkInterfaces) == 1 &&
+		(len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 ||
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) != nil) {
+
+		ipv4CIDR := ""
+		if len(instance.NetworkInterfaces[0].AliasIpRanges) > 0 {
+			ipv4CIDR = instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange
+		}
+
+		ipv6Addr := ""
+		if addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]); addr != nil {
+			ipv6Addr = addr.String()
+		}
+
+		switch ca.stackType {
+		case stackIPv4:
+			cidrStrings = append(cidrStrings, ipv4CIDR)
+		case stackDualStack:
+			cidrStrings = append(cidrStrings, ipv4CIDR)
+			cidrStrings = append(cidrStrings, ipv6Addr)
+		case stackIPv6:
+			cidrStrings = append(cidrStrings, ipv6Addr)
```

Comment:
you can avoid the appends and do the assignment directly, this is less error prone

```suggestion
			cidrStrings = []string{ipv4CIDR}
		case stackDualStack:
			cidrStrings = []string{ipv4CIDR,ipv6Addr}
		case stackIPv6:
			cidrStrings = []string{ipv6Addr}
```

Reply:
you have also to validate that if is ipv4 the cidr exists, as if is ipv6 only but there are not ipv6 addresses this will be an empty cidrString ... or is this validated later?

Reply:
Thanks for your offline explanations @aojea!

1. I was wrong when I assumed the `ca.cloud.GetIPV6Address` function makes an external call. In fact it only parses the network interface struct. This was slightly confusing to me, but maybe totally fine in fact. I assumed `cloud` would be a GCE client, allowing RPCs towards the Compute API. In fact this is a wrapper that sometimes would make RPC calls, but sometimes would only process the data locally. It seems this is something I just need to get used to, so nothing wrong with the code.

2. I looked into how I could test the case of a mismatched configuration - service CIDRs not aligned with the network interface's configuration, e. g. service CIDRs suggesting the cluster is single-stack IPv6, while the network interface in fact only has IPv4 addresses. The good thing is that it will work the right way, and if there are no pod CIDRs to put on a node object, the code will generate an error in https://github.com/kubernetes/cloud-provider-gcp/blob/master/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L385.
There is no good way to test it, though. Unfortunately in a test the cloud CIDR allocator is initialized directly (https://github.com/kubernetes/cloud-provider-gcp/blob/master/pkg/controller/nodeipam/ipam/cloud_cidr_allocator_test.go#L1323) instead of using the function `NewCloudCIDRAllocator()` (https://github.com/kubernetes/cloud-provider-gcp/blob/master/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L86). Because of that the tests skip the logic in the function, including the code parsing service CIDRs I'm just adding to the `NewCloudCIDRAllocator()` function as part of this PR. I think we should try to restructure the tests so they use the `NewCloudCIDRAllocator()` function - but maybe not as part of this PR? WDYT @aojea ?